README.md
jfinal cms
------------------------

> 1. jfinal cms，采用了简洁强大的JFinal作为web框架，模板引擎用的是beetl，数据库用mysql，前端bootstrap框架。 
> 2. 后台模块包含：栏目管理，栏目公告，栏目滚动图片，文章管理，回复管理，意见反馈，我的相册，相册管理，图片管理，专辑管理，视频管理，缓存更新，友情链接，访问统计，联系人管理，模板管理，组织机构管理，用户管理，角色管理，菜单管理，参数配置，数据字典管理。
> 3. 后端模板支持：bootstrap默认样式、bootstrap黑色样式和flat-ui样式
> 4. 前端模板支持：默认内容发布、官网模板、图片模板和视频模板
> 5. jfinal cms交流群：568909653。 文档见doc/jfinal cms文档.docx

* 管理地址：http://${ip:port}/${project_name}/admin
* 测试账号: admin/admin123 或 test/123456

平台部署和配置说明
------------------------

> 1. 下载项目代码，安装jdk、maven、mysql。
> 2. 在项目目录下运行mvn install，提示BUILD SUCCESS即可。
> 3. 创建mysql用户和数据库，运行/jfinal_cms/sql下对应jfinal_cms_v4.sql。
> 4. 数据库配置文件：/jfinal_cms/src/main/resources/conf/db.properties
> 5. 如需要oauth2的，设置src/conf/oauth.properties
> 6. 运行：mvn tomcat:run
> 7. 系统默认采用单站点模式，各个站点可以在“其他管理”下“站点管理”菜单方便的切换。
> 8. 如果使用多站点，可以在“系统管理”下“多站点标示”菜单中，将“多站点标示”项目修改为true。
> 9. 多站点需要设置各个站点对应的域名，通过域名解析到不同的站点模板。


项目源码地址：
------------------------

github地址：https://github.com/jflyfox/jfinal_cms

码云地址：https://git.oschina.net/flyfox/jfinal_cms

API Clinet 项目源码地址：
------------------------

github地址：https://github.com/jflyfox/jfinal_cms_api_client

码云地址：https://git.oschina.net/jflyfox/jfinal_cms_api_client

其他说明
------------------------

两个依赖项目jflyfox_base，jflyfox_jfinal。前者是String，Number,Date等公共方法封装，后者是对jfinal和一些公用方法的简单封装

源码地址：
> http://git.oschina.net/flyfox/jflyfox_jfinal 

> http://git.oschina.net/flyfox/jflyfox_base

演示效果截图
------------------------


#### 网站CMS地址：[http://mtg.jflyfox.com/](http://mtg.jflyfox.com/) ####
![网站](http://static.oschina.net/uploads/img/201601/21022316_Nk5M.gif "jfinal cms")

#### 网站官网模板：[http://website.jflyfox.com/](http://website.jflyfox.com/) ####
![官网](http://static.oschina.net/uploads/img/201601/21022316_XkxY.gif "jfinal cms")

#### 博客模板模板：[http://blog.jflyfox.com/](http://blog.jflyfox.com) ####
![官网](http://static.oschina.net/uploads/space/2016/0622/002206_Rla0_166354.jpg "jfinal cms")

#### 相册管理模板：[http://photo.jflyfox.com/](http://photo.jflyfox.com/) ####
![官网](http://static.oschina.net/uploads/space/2016/0306/144741_ldOJ_166354.gif "jfinal cms")

#### 视频管理模板：[http://video.jflyfox.com/](http://video.jflyfox.com/) ####
![官网](http://static.oschina.net/uploads/space/2016/0306/144754_FXhR_166354.gif "jfinal cms")

#### 后台页面主题： ####
![后台管理](http://static.oschina.net/uploads/img/201601/28091447_rQtD.gif "jfinal cms")

鸣谢
------------------------

 1. [JFinal](http://www.oschina.net/p/jfinal)
 2. [beetl](http://ibeetl.com/community/)
 3. [oschina](http://www.oschina.net/)

support
------------------------

项目的发展，离不开大家得支持~！~

![jflyfox](http://blog.jflyfox.com/static/images/common/pay_weixin.jpg "Open source support(alipay)")

![jflyfox](http://blog.jflyfox.com/static/images/common/pay_alipay.jpg "Open source support(weixin)")



ChangeLog.md
Change Log 更新说明
------------------------------

## 2017-06-01
1. 修复多站点和单站点切换导致的异常。

## 2017-05-26 jfinal cms4.6.0
1. jflyfox_jfinal升级为4.5.0
2. 重构站点逻辑代码，让后台站点配置与前台站点配置分离；
3. delete、insert、update默认不保存日志信息；如果需要可以默认保存可以恢复BaseProjectModel类；
4. 规范SessionUser和SysUser属性，避免相同属性不同写法问题；
5. 解决ueditor和umeditor可能会报重复注册controller错误；
6. 解决IE提交数据可能保存两条数据问题；

## 2017-05-02
1. 将密码由des改为3des加密。
2. 修复登陆cookie异常，导致不能记住密码问题。
3. 加入用户密码重置功能。
4. 加入api秘钥展示功能。

## 2017-04-17
1. 修复IE11重复提交数据问题。

### 2017-04-13
1. 添加jfinal cms api文档

### 2017-04-05 jfinal cms4.5.0
1. 添加文件恢复机制。
2. 加入阿里云OSS备份和恢复处理。
3. 完善文档，细节优化并加入了站点配置和文件备份配置说明。

### 2017-04-04
1. 完成上传备份机制。
2. ueditor加入上传备份及站点机制。
3. 优化上传路径处理机制。

### 2017-03-23
1. 组织机构根目录展示问题修复。

### 2017-03-22    jfinal cms4.4.0
1. API接口全面完善，加入参数加密和认证功能
2. 可以通过配置切换加密和认证功能开启
3. 修复fst导致第三方Oauth2登陆异常问题

### 2017-03-14
1. 加入API用户类型，组织机构树结构调整

### 2017-03-05
1. 后台登陆加入md5传输，防止网络拦截。

### 2017-02-08    jfinal cms4.3.0
1. 缓存加入redis支持
2. 第三方核心包升级：jfinal v3.0、beetl v2.7.11

### 2017-01-22    jfinal cms4.2.0
1. 优化论坛模板数据
2. 模板发布功能优化

### 2017-01-20   
1. 加入论坛模板

### 2017-01-07   
1. 完善博客样式
2. 加入参数配置菜单更新sql
3. 完善权限细节

### 2017-01-01    jfinal cms4.1.5
1. 后台加入新的主题flat。

### 2016-12-31
1. 优化站点配置后台以及拦截逻辑。
2. 将站点设置加入到参数管理中。
3. 默认为单站点，可以方便的通过站点管理切换。

### 2016-12-17    jfinal cms4.1.4
1. 加入系统参数配置模块。
2. api配置迁移到系统参数配置中。
3. 修复登陆后可以跳过菜单限制，加入菜单权限拦截。

### 2016-11-15
1. api接口调用加入了登陆验证以及验证开关。

### 2016-10-03   jfinal cms4.1.3
1. 完善api功能。
2. 加入api接口文档。

### 2016-10-02
1. 加入api功能。

### 2016-09-21   jfinal cms4.1.2
1. 完善初始化数据。
2. 完善blog模板。
3. 修改默认模板为blog。

### 2016-09-20
1. 加入blog模板。

### 2016-08-01  jfinal cms4.1.1
1. 多站点key冲突问题。

### 2016-07-31
1. 组织机构优化，加入编码和上级组织机构。
2. 组织机构加入树形展示。


### 2016-06-23
1. 修复异常数据问题。
2. 修复sql多版本兼容问题。
3. 修复多站点UrlKey一直导致的不可预知bug。


### 2016-06-21  jfinal cms4.1.0
1. jfinal cms 加入project项目模板。

### 2016-06-12
1. jfinal cms项目文档完善。

### 2016-06-10
1. 加入jfinal cms项目文档。

### 2016-06-01
1. 修复项目路径为根目录时，模板加载异常问题。

### 2016-05-29
1. 修复标签查询图片展示异常问题

### 2016-05-11  jfinal cms4.0.0
1. 升级为4.0.0

### 2016-05-04
1. 友情链接加入多站点支持。
2. 用户属性加入站点配置，避免再次登录重复选择站点问题。
3. 用户登录限制优化，加入用户类型和创建站点设置。
4. 加入菜单权限拦截器，能够阻止大部分异常操作。
5. 支持目录模板路径设置。

### 2016-04-27
1. 修改上传路径~支持多站点目录结构。

### 2016-04-25
1. 多站点和单站点支持优化。

### 2016-04-18
1. 加入ue编辑器
2. 上传目录修改
3. 高亮功能
4. oauth跳转问题修复

### 2016-04-15
1. 站点管理持续优化

### 2016-03-19  jfinal cms3.1.0
1. 升级为3.1.0
2. 修复cookie异常，持续报错问题。
3. 加入文章直接跳转到第三方功能。
4. 实现前后台登陆隔离。
5. 后台菜单引用路径重构。
6. 加入文章审核功能，可以配置是否需要审核。
7. 加入文章复制功能。

### 2016-03-13
1. 加入栏目和文章树结构展示

### 2016-03-06  jfinal cms3.0.0
1. 支持模板文件在线编辑，升级为3.0.0

### 2016-03-06
1. 优化模板管理
2. 加入主题切换。
3. 加入修改文件备份处理。

### 2016-03-04
1. 加入模板管理模块，引入文件管理。
2. filemanager来自github，再此基础上进行了路径优化，主题切换支持，中文支持，以及replaceFile，addFile接口完善。

### 2016-02-24
1. jflyfox_base修改为3.1修复中文和空格目录问题
2. 解决pom中mvn tomcat:run 目录问题

### 2016-02-20 jfinal cms2.9.0
1. 重构为maven项目，升级为2.9.0

### 2016-02-20 jfinal cms2.8.0
1. 升级为2.8.0

### 2016-02-20
1. 完善视频模板。
2. 完善视频模板数据。

### 2016-02-19
1. 加入后台视频专辑和视频模块。
2. 加入前端视频模板。
3. 重构前端图片模板路径。

### 2016-02-14 jfinal cms2.7.0
1. 升级2.7.0。

### 2016-02-10
1. 完善相册管理，加入我的相册。
2. 加入相册标签编辑。
3. 加入相册管理前台模板。
4. 图片查看页面优化。
5. 相册管理模板加入分页展示。

### 2016-02-5
1. 加入相册管理和图片管理。

### 2016-02-03
1. 加入后台首页面。
2. 登录成功后，跳转到首页面。

### 2016-01-31 jfinal cms2.6.0
1. 加入前台栏目公告。
2. 优化

### 2016-01-29
1. 意见反馈优化
2. 优化菜单选择，重新整理菜单

### 2016-01-28
1. 加入意见反馈，栏目滚屏，栏目公告三个模块后台代码。
2. 迁移前台首页滚屏数据。
3. 加入前台意见反馈。

### 2016-01-28 jfinal cms2.5.0
1. 升级为2.5.0。
12. 后台加入主题切换，bootstrap默认，黑色，flat-ui三个主题。

### 2016-01-20 jfinal cms2.4.0
1. 升级为2.4.0。
2. 项目前台页面全面重构，jfinal升级为2.2。
3. 前台加入模板扩展功能，通过beetl functions获取后台数据。
原有jfinal cms模板不变，加入website官网模板。


src/main/java/com/jflyfox/component/interceptor/CommonInterceptor.java
package com.jflyfox.component.interceptor;

import com.jfinal.aop.Interceptor;
import com.jfinal.aop.Invocation;
import com.jfinal.core.Controller;
import com.jflyfox.component.util.JFlyFoxUtils;
import com.jflyfox.jfinal.base.SessionUser;
import com.jflyfox.jfinal.component.util.Attr;
import com.jflyfox.modules.admin.site.SessionSite;
import com.jflyfox.modules.admin.site.SiteConstant;
import com.jflyfox.modules.admin.site.SiteService;
import com.jflyfox.modules.admin.site.TbSite;
import com.jflyfox.system.config.ConfigCache;

/**
 * 公共属性拦截器
 * 
 * @author flyfox 2014-2-11
 */
public class CommonInterceptor implements Interceptor {

	private SiteService siteSvc = new SiteService();

	@SuppressWarnings("rawtypes")
	public void intercept(Invocation ai) {

		Controller controller = ai.getController();

		// 请求路径
		String tmpPath = ai.getActionKey();
		tmpPath = JFlyFoxUtils.handlerPath(tmpPath);
		boolean isBack = JFlyFoxUtils.isBack(tmpPath);

		if (isBack) { // 后台
			SessionUser user = controller.getSessionAttr(Attr.SESSION_NAME);
			if (user == null) {
				String title = ConfigCache.getValue("SYSTEM.TITLE");
				// 设置公共属性
				controller.setAttr(JFlyFoxUtils.WEBSITE_TITLE, title);
				controller.setAttr(JFlyFoxUtils.TITLE_ATTR, title);
				controller.setAttr(JFlyFoxUtils.KEYWORDS_ATTR, title);
				controller.setAttr(JFlyFoxUtils.DESCRIPTION_ATTR, title);
			} else {
				TbSite site = siteSvc.getSite(user.getBackSiteId());
				if(null == site){
					site = siteSvc.getDefaultSite();
				}
				controller.setAttr(JFlyFoxUtils.WEBSITE_TITLE, site.getSiteTitle());
				controller.setAttr(JFlyFoxUtils.TITLE_ATTR, site.getSiteTitle());
				controller.setAttr(JFlyFoxUtils.KEYWORDS_ATTR, site.getSiteTitle());
				controller.setAttr(JFlyFoxUtils.DESCRIPTION_ATTR, site.getSiteTitle());
			}
		} else {
			SessionSite sessionSite = controller.getSessionAttr(SiteConstant.getSessionSite());
			if (sessionSite == null) {
				String title = ConfigCache.getValue("SYSTEM.TITLE");
				// 设置公共属性
				controller.setAttr(JFlyFoxUtils.WEBSITE_TITLE, title);
				controller.setAttr(JFlyFoxUtils.TITLE_ATTR, title);
				controller.setAttr(JFlyFoxUtils.KEYWORDS_ATTR, title);
				controller.setAttr(JFlyFoxUtils.DESCRIPTION_ATTR, title);
			} else if (sessionSite.getModel() != null) {
				TbSite site = sessionSite.getModel();
				controller.setAttr(JFlyFoxUtils.WEBSITE_TITLE, site.getSiteTitle());
				controller.setAttr(JFlyFoxUtils.TITLE_ATTR, site.getSiteTitle());
				controller.setAttr(JFlyFoxUtils.KEYWORDS_ATTR, site.getSiteTitle());
				controller.setAttr(JFlyFoxUtils.DESCRIPTION_ATTR, site.getSiteTitle());
			}
		}

		ai.invoke();
	}
}


src/main/java/com/jflyfox/modules/front/template/TemplateImageService.java
package com.jflyfox.modules.front.template;

import com.jfinal.plugin.activerecord.Page;
import com.jflyfox.jfinal.base.BaseService;
import com.jflyfox.jfinal.base.Paginator;
import com.jflyfox.modules.admin.image.model.TbImage;
import com.jflyfox.modules.admin.image.model.TbImageAlbum;
import com.jflyfox.modules.front.service.FrontImageService;
import com.jflyfox.util.extend.RandomStrUtils;

import java.util.List;

/**
 * 模板方法接口
 * 
 * 2016年1月18日 下午6:05:54 flyfox 369191470@qq.com
 */
public class TemplateImageService extends BaseService {

	private final static FrontImageService service = new FrontImageService();

	public String randomAlbumId() {
		return RandomStrUtils.randomAlphanumeric(5);
	}

	public List<TbImageAlbum> albums() {
		return service.getAlbumList();
	}

	public List<TbImageAlbum> albmsByParentId(int parentId){
		return  service.getAlbumListByParentId(parentId);
	}
	public TbImageAlbum album(Integer albumId) {
		return service.getAlbum(albumId);
	}

	public Page<TbImage> images(int pageNo, int pageSize) {
		return service.getImage(new Paginator(pageNo, pageSize));
	}

	public Page<TbImage> images(int pageNo, int pageSize, int albumId) {
		return service.getImage(new Paginator(pageNo, pageSize), albumId);
	}

	public TbImage image(Integer iamgeId) {
		return service.getImage(iamgeId);
	}

	public Page<TbImage> recommendImages(int pageNo, int pageSize) {
		return service.getRecommendImages(new Paginator(pageNo, pageSize));
	}
}

src/main/webapp/static/component/ueditor/ueditor.config.js
/**
 * ueditor完整配置项
 * 可以在这里配置整个编辑器的特性
 */
/**************************提示********************************
 * 所有被注释的配置项均为UEditor默认值。
 * 修改默认配置请首先确保已经完全明确该参数的真实用途。
 * 主要有两种修改方案，一种是取消此处注释，然后修改成对应参数；另一种是在实例化编辑器时传入对应参数。
 * 当升级编辑器时，可直接使用旧版配置文件替换新版配置文件,不用担心旧版配置文件中因缺少新功能所需的参数而导致脚本报错。
 **************************提示********************************/

(function () {

    /**
     * 编辑器资源文件根路径。它所表示的含义是：以编辑器实例化页面为当前路径，指向编辑器资源文件（即dialog等文件夹）的路径。
     * 鉴于很多同学在使用编辑器的时候出现的种种路径问题，此处强烈建议大家使用"相对于网站根目录的相对路径"进行配置。
     * "相对于网站根目录的相对路径"也就是以斜杠开头的形如"/myProject/ueditor/"这样的路径。
     * 如果站点中有多个不在同一层级的页面需要实例化编辑器，且引用了同一UEditor的时候，此处的URL可能不适用于每个页面的编辑器。
     * 因此，UEditor提供了针对不同页面的编辑器可单独配置的根路径，具体来说，在需要实例化编辑器的页面最顶部写上如下代码即可。当然，需要令此处的URL等于对应的配置。
     * window.UEDITOR_HOME_URL = "/xxxx/xxxx/";
     */
    var URL = window.UEDITOR_HOME_URL || getUEBasePath();

    /**
     * 配置项主体。注意，此处所有涉及到路径的配置别遗漏URL变量。
     */
    window.UEDITOR_CONFIG = {

        //为编辑器实例添加一个路径，这个不能被注释
        UEDITOR_HOME_URL: URL

        // 服务器统一请求接口路径
        , serverUrl: URL+"../../../ueditor/"

        //工具栏上的所有的功能按钮和下拉框，可以在new编辑器的实例时选择自己需要的从新定义
        , toolbars: [[
            'fullscreen', 'source', '|', 'undo', 'redo', '|',
            'bold', 'italic', 'underline', 'fontborder', 'strikethrough', 'superscript', 'subscript', 'removeformat', 'formatmatch', 'autotypeset',  '|',
            'forecolor', 'backcolor', 'insertorderedlist', 'insertunorderedlist', '|',
            'paragraph', 'fontfamily', 'fontsize', '|',
            'justifyleft', 'justifycenter', 'justifyright', 'justifyjustify', '|',
            'link', 'unlink', '|', 'imagenone', 'imageleft', 'imageright', 'imagecenter', '|',
            'simpleupload', 'insertimage', 'emotion', 'scrawl', 'insertvideo', 'music', 'attachment', 'map', 'insertframe', 'insertcode', 'pagebreak', 'template', '|',
            'horizontal', 'date', 'time', 'spechars', '|',
            'inserttable', 'print', 'preview', 'searchreplace'
        ]]
	    ,'insertcode':{
	    	'java':'Java',
	    	'html':'Html',
            'css':'Css',
            'js':'Javascript',
            'xml':'Xml',
            'sql':'Sql',
            'bash':'Bash/Shell',
            'jfx':'JavaFx',
            'cpp':'C/C++',
            'php':'Php'
	    }
        //当鼠标放在工具栏上时显示的tooltip提示,留空支持自动多语言配置，否则以配置值为准
        //,labelMap:{
        //    'anchor':'', 'undo':''
        //}

        //语言配置项,默认是zh-cn。有需要的话也可以使用如下这样的方式来自动多语言切换，当然，前提条件是lang文件夹下存在对应的语言文件：
        //lang值也可以通过自动获取 (navigator.language||navigator.browserLanguage ||navigator.userLanguage).toLowerCase()
        //,lang:"zh-cn"
        //,langPath:URL +"lang/"

        //主题配置项,默认是default。有需要的话也可以使用如下这样的方式来自动多主题切换，当然，前提条件是themes文件夹下存在对应的主题文件：
        //现有如下皮肤:default
        //,theme:'default'
        //,themePath:URL +"themes/"

        //,zIndex : 900     //编辑器层级的基数,默认是900

        //针对getAllHtml方法，会在对应的head标签中增加该编码设置。
        //,charset:"utf-8"

        //若实例化编辑器的页面手动修改的domain，此处需要设置为true
        //,customDomain:false

        //常用配置项目
        //,isShow : true    //默认显示编辑器

        //,textarea:'editorValue' // 提交表单时，服务器获取编辑器提交内容的所用的参数，多实例时可以给容器name属性，会将name给定的值最为每个实例的键值，不用每次实例化的时候都设置这个值

        //,initialContent:'欢迎使用ueditor!'    //初始化编辑器的内容,也可以通过textarea/script给值，看官网例子

        //,autoClearinitialContent:true //是否自动清除编辑器初始内容，注意：如果focus属性设置为true,这个也为真，那么编辑器一上来就会触发导致初始化的内容看不到了

        //,focus:false //初始化时，是否让编辑器获得焦点true或false

        //如果自定义，最好给p标签如下的行高，要不输入中文时，会有跳动感
        //,initialStyle:'p{line-height:1em}'//编辑器层级的基数,可以用来改变字体等

        //,iframeCssUrl: URL + '/themes/iframe.css' //给编辑器内部引入一个css文件

        //indentValue
        //首行缩进距离,默认是2em
        //,indentValue:'2em'

        //,initialFrameWidth:1000  //初始化编辑器宽度,默认1000
        //,initialFrameHeight:320  //初始化编辑器高度,默认320

        //,readonly : false //编辑器初始化结束后,编辑区域是否是只读的，默认是false

        //,autoClearEmptyNode : true //getContent时，是否删除空的inlineElement节点（包括嵌套的情况）

        //启用自动保存
        //,enableAutoSave: true
        //自动保存间隔时间， 单位ms
        //,saveInterval: 500

        //,fullscreen : false //是否开启初始化时即全屏，默认关闭

        //,imagePopup:true      //图片操作的浮层开关，默认打开

        //,autoSyncData:true //自动同步编辑器要提交的数据
        //,emotionLocalization:false //是否开启表情本地化，默认关闭。若要开启请确保emotion文件夹下包含官网提供的images表情文件夹

        //粘贴只保留标签，去除标签所有属性
        //,retainOnlyLabelPasted: false

        //,pasteplain:false  //是否默认为纯文本粘贴。false为不使用纯文本粘贴，true为使用纯文本粘贴
        //纯文本粘贴模式下的过滤规则
        //'filterTxtRules' : function(){
        //    function transP(node){
        //        node.tagName = 'p';
        //        node.setStyle();
        //    }
        //    return {
        //        //直接删除及其字节点内容
        //        '-' : 'script style object iframe embed input select',
        //        'p': {$:{}},
        //        'br':{$:{}},
        //        'div':{'$':{}},
        //        'li':{'$':{}},
        //        'caption':transP,
        //        'th':transP,
        //        'tr':transP,
        //        'h1':transP,'h2':transP,'h3':transP,'h4':transP,'h5':transP,'h6':transP,
        //        'td':function(node){
        //            //没有内容的td直接删掉
        //            var txt = !!node.innerText();
        //            if(txt){
        //                node.parentNode.insertAfter(UE.uNode.createText(' &nbsp; &nbsp;'),node);
        //            }
        //            node.parentNode.removeChild(node,node.innerText())
        //        }
        //    }
        //}()

        //,allHtmlEnabled:false //提交到后台的数据是否包含整个html字符串

        //insertorderedlist
        //有序列表的下拉配置,值留空时支持多语言自动识别，若配置值，则以此值为准
        //,'insertorderedlist':{
        //      //自定的样式
        //        'num':'1,2,3...',
        //        'num1':'1),2),3)...',
        //        'num2':'(1),(2),(3)...',
        //        'cn':'一,二,三....',
        //        'cn1':'一),二),三)....',
        //        'cn2':'(一),(二),(三)....',
        //     //系统自带
        //     'decimal' : '' ,         //'1,2,3...'
        //     'lower-alpha' : '' ,    // 'a,b,c...'
        //     'lower-roman' : '' ,    //'i,ii,iii...'
        //     'upper-alpha' : '' , lang   //'A,B,C'
        //     'upper-roman' : ''      //'I,II,III...'
        //}

        //insertunorderedlist
        //无序列表的下拉配置，值留空时支持多语言自动识别，若配置值，则以此值为准
        //,insertunorderedlist : { //自定的样式
        //    'dash' :'— 破折号', //-破折号
        //    'dot':' 。 小圆圈', //系统自带
        //    'circle' : '',  // '○ 小圆圈'
        //    'disc' : '',    // '● 小圆点'
        //    'square' : ''   //'■ 小方块'
        //}
        //,listDefaultPaddingLeft : '30'//默认的左边缩进的基数倍
        //,listiconpath : 'http://bs.baidu.com/listicon/'//自定义标号的路径
        //,maxListLevel : 3 //限制可以tab的级数, 设置-1为不限制

        //,autoTransWordToList:false  //禁止word中粘贴进来的列表自动变成列表标签

        //fontfamily
        //字体设置 label留空支持多语言自动切换，若配置，则以配置值为准
        //,'fontfamily':[
        //    { label:'',name:'songti',val:'宋体,SimSun'},
        //    { label:'',name:'kaiti',val:'楷体,楷体_GB2312, SimKai'},
        //    { label:'',name:'yahei',val:'微软雅黑,Microsoft YaHei'},
        //    { label:'',name:'heiti',val:'黑体, SimHei'},
        //    { label:'',name:'lishu',val:'隶书, SimLi'},
        //    { label:'',name:'andaleMono',val:'andale mono'},
        //    { label:'',name:'arial',val:'arial, helvetica,sans-serif'},
        //    { label:'',name:'arialBlack',val:'arial black,avant garde'},
        //    { label:'',name:'comicSansMs',val:'comic sans ms'},
        //    { label:'',name:'impact',val:'impact,chicago'},
        //    { label:'',name:'timesNewRoman',val:'times new roman'}
        //]

        //fontsize
        //字号
        //,'fontsize':[10, 11, 12, 14, 16, 18, 20, 24, 36]

        //paragraph
        //段落格式 值留空时支持多语言自动识别，若配置，则以配置值为准
        //,'paragraph':{'p':'', 'h1':'', 'h2':'', 'h3':'', 'h4':'', 'h5':'', 'h6':''}

        //rowspacingtop
        //段间距 值和显示的名字相同
        //,'rowspacingtop':['5', '10', '15', '20', '25']

        //rowspacingBottom
        //段间距 值和显示的名字相同
        //,'rowspacingbottom':['5', '10', '15', '20', '25']

        //lineheight
        //行内间距 值和显示的名字相同
        //,'lineheight':['1', '1.5','1.75','2', '3', '4', '5']

        //customstyle
        //自定义样式，不支持国际化，此处配置值即可最后显示值
        //block的元素是依据设置段落的逻辑设置的，inline的元素依据BIU的逻辑设置
        //尽量使用一些常用的标签
        //参数说明
        //tag 使用的标签名字
        //label 显示的名字也是用来标识不同类型的标识符，注意这个值每个要不同，
        //style 添加的样式
        //每一个对象就是一个自定义的样式
        //,'customstyle':[
        //    {tag:'h1', name:'tc', label:'', style:'border-bottom:#ccc 2px solid;padding:0 4px 0 0;text-align:center;margin:0 0 20px 0;'},
        //    {tag:'h1', name:'tl',label:'', style:'border-bottom:#ccc 2px solid;padding:0 4px 0 0;margin:0 0 10px 0;'},
        //    {tag:'span',name:'im', label:'', style:'font-style:italic;font-weight:bold'},
        //    {tag:'span',name:'hi', label:'', style:'font-style:italic;font-weight:bold;color:rgb(51, 153, 204)'}
        //]

        //打开右键菜单功能
        //,enableContextMenu: true
        //右键菜单的内容，可以参考plugins/contextmenu.js里边的默认菜单的例子，label留空支持国际化，否则以此配置为准
        //,contextMenu:[
        //    {
        //        label:'',       //显示的名称
        //        cmdName:'selectall',//执行的command命令，当点击这个右键菜单时
        //        //exec可选，有了exec就会在点击时执行这个function，优先级高于cmdName
        //        exec:function () {
        //            //this是当前编辑器的实例
        //            //this.ui._dialogs['inserttableDialog'].open();
        //        }
        //    }
        //]

        //快捷菜单
        //,shortcutMenu:["fontfamily", "fontsize", "bold", "italic", "underline", "forecolor", "backcolor", "insertorderedlist", "insertunorderedlist"]

        //elementPathEnabled
        //是否启用元素路径，默认是显示
        //,elementPathEnabled : true

        //wordCount
        //,wordCount:true          //是否开启字数统计
        //,maximumWords:10000       //允许的最大字符数
        //字数统计提示，{#count}代表当前字数，{#leave}代表还可以输入多少字符数,留空支持多语言自动切换，否则按此配置显示
        //,wordCountMsg:''   //当前已输入 {#count} 个字符，您还可以输入{#leave} 个字符
        //超出字数限制提示  留空支持多语言自动切换，否则按此配置显示
        //,wordOverFlowMsg:''    //<span style="color:red;">你输入的字符个数已经超出最大允许值，服务器可能会拒绝保存！</span>

        //tab
        //点击tab键时移动的距离,tabSize倍数，tabNode什么字符做为单位
        //,tabSize:4
        //,tabNode:'&nbsp;'

        //removeFormat
        //清除格式时可以删除的标签和属性
        //removeForamtTags标签
        //,removeFormatTags:'b,big,code,del,dfn,em,font,i,ins,kbd,q,samp,small,span,strike,strong,sub,sup,tt,u,var'
        //removeFormatAttributes属性
        //,removeFormatAttributes:'class,style,lang,width,height,align,hspace,valign'

        //undo
        //可以最多回退的次数,默认20
        //,maxUndoCount:20
        //当输入的字符数超过该值时，保存一次现场
        //,maxInputCount:1

        //autoHeightEnabled
        // 是否自动长高,默认true
        ,autoHeightEnabled:false

        //scaleEnabled
        //是否可以拉伸长高,默认true(当开启时，自动长高失效)
        //,scaleEnabled:false
        //,minFrameWidth:800    //编辑器拖动时最小宽度,默认800
        //,minFrameHeight:220  //编辑器拖动时最小高度,默认220

        //autoFloatEnabled
        //是否保持toolbar的位置不动,默认true
        //,autoFloatEnabled:true
        //浮动时工具栏距离浏览器顶部的高度，用于某些具有固定头部的页面
        //,topOffset:30
        //编辑器底部距离工具栏高度(如果参数大于等于编辑器高度，则设置无效)
        //,toolbarTopOffset:400

        //pageBreakTag
        //分页标识符,默认是_ueditor_page_break_tag_
        //,pageBreakTag:'_ueditor_page_break_tag_'

        //autotypeset
        //自动排版参数
        //,autotypeset: {
        //    mergeEmptyline: true,           //合并空行
        //    removeClass: true,              //去掉冗余的class
        //    removeEmptyline: false,         //去掉空行
        //    textAlign:"left",               //段落的排版方式，可以是 left,right,center,justify 去掉这个属性表示不执行排版
        //    imageBlockLine: 'center',       //图片的浮动方式，独占一行剧中,左右浮动，默认: center,left,right,none 去掉这个属性表示不执行排版
        //    pasteFilter: false,             //根据规则过滤没事粘贴进来的内容
        //    clearFontSize: false,           //去掉所有的内嵌字号，使用编辑器默认的字号
        //    clearFontFamily: false,         //去掉所有的内嵌字体，使用编辑器默认的字体
        //    removeEmptyNode: false,         // 去掉空节点
        //    //可以去掉的标签
        //    removeTagNames: {标签名字:1},
        //    indent: false,                  // 行首缩进
        //    indentValue : '2em',            //行首缩进的大小
        //    bdc2sb: false,
        //    tobdc: false
        //}

        //tableDragable
        //表格是否可以拖拽
        //,tableDragable: true

        //,disabledTableInTable:true  //禁止表格嵌套

        //sourceEditor
        //源码的查看方式,codemirror 是代码高亮，textarea是文本框,默认是codemirror
        //注意默认codemirror只能在ie8+和非ie中使用
        ,sourceEditor:"codemirror"
        //如果sourceEditor是codemirror，还用配置一下两个参数
        //codeMirrorJsUrl js加载的路径，默认是 URL + "third-party/codemirror/codemirror.js"
        //,codeMirrorJsUrl:URL + "third-party/codemirror/codemirror.js"
        //codeMirrorCssUrl css加载的路径，默认是 URL + "third-party/codemirror/codemirror.css"
        //,codeMirrorCssUrl:URL + "third-party/codemirror/codemirror.css"
        //编辑器初始化完成后是否进入源码模式，默认为否。
        //,sourceEditorFirst:false

        //iframeUrlMap
        //dialog内容的路径 ～会被替换成URL,垓属性一旦打开，将覆盖所有的dialog的默认路径
        //,iframeUrlMap:{
        //    'anchor':'~/dialogs/anchor/anchor.html',
        //}

        //webAppKey 百度应用的APIkey，每个站长必须首先去百度官网注册一个key后方能正常使用app功能，注册介绍，http://app.baidu.com/static/cms/getapikey.html
        //, webAppKey: ""
    };

    function getUEBasePath(docUrl, confUrl) {

        return getBasePath(docUrl || self.document.URL || self.location.href, confUrl || getConfigFilePath());

    }

    function getConfigFilePath() {

        var configPath = document.getElementsByTagName('script');

        return configPath[ configPath.length - 1 ].src;

    }

    function getBasePath(docUrl, confUrl) {

        var basePath = confUrl;


        if (/^(\/|\\\\)/.test(confUrl)) {

            basePath = /^.+?\w(\/|\\\\)/.exec(docUrl)[0] + confUrl.replace(/^(\/|\\\\)/, '');

        } else if (!/^[a-z]+:/i.test(confUrl)) {

            docUrl = docUrl.split("#")[0].split("?")[0].replace(/[^\\\/]+$/, '');

            basePath = docUrl + "" + confUrl;

        }

        return optimizationPath(basePath);

    }

    function optimizationPath(path) {

        var protocol = /^[a-z]+:\/\//.exec(path)[ 0 ],
            tmp = null,
            res = [];

        path = path.replace(protocol, "").split("?")[0].split("#")[0];

        path = path.replace(/\\/g, '/').split(/\//);

        path[ path.length - 1 ] = "";

        while (path.length) {

            if (( tmp = path.shift() ) === "..") {
                res.pop();
            } else if (tmp !== ".") {
                res.push(tmp);
            }

        }

        return protocol + res.join("/");

    }

    window.UE = {
        getUEBasePath: getUEBasePath
    };

})();
